[WIP] GraphQL+ parser fuzzing

pborzenkov opened the pull request:
The PR adds bits required to fuzz GraphQL+ parser along with test cases found by go-fuzz that cause parser crash.
The PR is incomplete as it doesn't yet fix all the identified problems.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/1565)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dgraph-io/dgraph?pullRequest=1565) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dgraph-io/dgraph?pullRequest=1565) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dgraph-io/dgraph?pullRequest=1565) it.</sub>

pborzenkov commented:
Added patches fix all crashes found by 10 minutes fuzzing session.
But this is definitely incomplete and the parser needs to be fuzzed a lot more.

pawanrawal commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks like a great change @pborzenkov. We could even use this for our query tests.

---

Review status: 0 of 211 files reviewed at latest revision, 3 unresolved discussions.

---

*[gql/parser_test.go, line 3998 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUNqwuRtsZOQzQ0ypX:-KvUNqwuRtsZOQzQ0ypY:b-umtpqd) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/parser_test.go#L3998)):*
> ```Go
> // We don't care here about return value, only about correct handling of
> // incorrect input.
> func TestParserFuzz(t *testing.T) {
> ```

Was this test generated by go-fuzz?

---

*[gql/fuzz-data/corpus/03f773418fcce70c93ad305c9bb36decddfa928a, line 1 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUOeI7px95vkEIhnO3:-KvUOeI7px95vkEIhnO4:b-oy1in2) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/fuzz-data/corpus/03f773418fcce70c93ad305c9bb36decddfa928a#L1)):*
> ```
> {e(func:uid(L),orderasc:val(c)){val(c)L as s{x as e}a as min(val(x))b as max(val(x))c as math(a+b)}}
> ```

I suppose these are generated by go-fuzz and correspond to the new failing cases?

---

*[gql/fuzz-data/corpus/test001.in, line 1 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUOWKOJYenjREEx9gJ:-KvUOWKOJYenjREEx9gK:bvcuqj0) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/fuzz-data/corpus/test001.in#L1)):*
> ```
> 
> ```

Did you extract all these manually from the tests in gql/parser_test.go?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-:-KvUOJPf2MXBIuoHqXbn:b-qzrf4y)*
<!-- Sent from Reviewable.io -->

pborzenkov commented:
Review status: 0 of 211 files reviewed at latest revision, 3 unresolved discussions.

---

*[gql/parser_test.go, line 3998 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUNqwuRtsZOQzQ0ypX:-KvWGef-AnEnxpgwazIa:b-wdjsq2) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/parser_test.go#L3998)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Was this test generated by go-fuzz?
</blockquote></details>

No. I added this test manually. Inputs are the ones the caused parser crash during the initial fuzzing. The idea is to be able to do regression testing on those inputs without running fuzzing every time (as it may require a lot of time). We can add new test cases here once they are discovered by go-fuzz.

---

*[gql/fuzz-data/corpus/03f773418fcce70c93ad305c9bb36decddfa928a, line 1 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUOeI7px95vkEIhnO3:-KvWHNtE-K4_NvgVK_XI:b-y6f7ch) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/fuzz-data/corpus/03f773418fcce70c93ad305c9bb36decddfa928a#L1)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I suppose these are generated by go-fuzz and correspond to the new failing cases?
</blockquote></details>

They haven't specifically caused problems during fuzzing, but rather increased coverage (go-fuzz is coverage guided). The idea is that preserving those corpus allow future go-fuzz sessions to proceed much quicker, as we provide example inputs that tell go-fuzz how to reach different code paths. Next fuzzing sessions will definitely find new code paths and will generate addition corpus. It's worth committing them as well. 

---

*[gql/fuzz-data/corpus/test001.in, line 1 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565#-KvUOWKOJYenjREEx9gJ:-KvWIL95wZ9et9GZsNiM:bbkbuli) ([raw file](https://github.com/dgraph-io/dgraph/blob/cfe1e7962c8af9cdff5fa4f93e5ec4617cf084d8/gql/fuzz-data/corpus/test001.in#L1)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Did you extract all these manually from the tests in gql/parser_test.go?
</blockquote></details>

No, not really :) I modified Parser function do dump its input into file and ran all TestParser* test cases. But yes, those file are all based on the original TestParser test cases.
They are not strictly required, as go-fuzz will eventually find proper inputs, but they speed up the fuzzing significantly. Basically, they provide some initial input that makes sense for Parser to go-fuzz, and it mutates it based on changes in coverage trying to find problematic inputs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1565)*
<!-- Sent from Reviewable.io -->
